design/volume-group-snapshot.md

```diff
@@ -0,0 +1,504 @@
+	for _, pvc := range pvcs {
+		names = append(names, pvc.Name)
+	}
+	return names
+}
+
+```
+
+- Add a new VolumeGroupSnapshot(VGS) BIA plugin
+```go
+// AppliesTo specifies that this plugin applies to VolumeGroupSnapshot objects.
+func (p *vgsBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumegroupsnapshots"},
+	}, nil
+}
+
+// Execute waits for all related VolumeSnapshot (VS) instances to be ready, then adds them
+// along with the VolumeGroupSnapshotClass as additional backup items.
+func (p *vgsBackupItemAction) Execute(
+	item runtime.Unstructured,
+	backup *velerov1api.Backup,
+) (
+	runtime.Unstructured,
+	[]velero.ResourceIdentifier,
+	string,
+	[]velero.ResourceIdentifier,
+	error,
+) {
+	p.log.Infof("Starting VGS backup plugin Execute for item: %s", item.GetName())
+
+	// Convert the unstructured object to a VolumeGroupSnapshot.
+	var vgs VolumeGroupSnapshot
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &vgs); err != nil {
+		return nil, nil, "", nil, errors.Wrap(err, "unable to convert item to VolumeGroupSnapshot")
+	}
+
+	// Wait until all related VolumeSnapshots are ready.
+	p.log.Infof("Waiting for all related VolumeSnapshots for VGS %s to be ready", vgs.Name)
+	ready, err := waitForGroupedVolumeSnapshots(&vgs, vgs.Namespace, p.log)
+	if err != nil {
+		return nil, nil, "", nil, errors.Wrap(err, "error waiting for grouped VolumeSnapshots")
+	}
+	if !ready {
+		p.log.Infof("Not all related VolumeSnapshots are ready for VGS %s", vgs.Name)
+		return item, nil, "", nil, nil
+	}
+
+	// Retrieve the related VolumeSnapshot instances.
+	groupedVS := getGroupedVolumeSnapshots(&vgs, vgs.Namespace)
+	var additionalItems []velero.ResourceIdentifier
+	for _, vs := range groupedVS {
+		additionalItems = append(additionalItems, velero.ResourceIdentifier{
+			GroupResource: schema.GroupResource{
+				Group:    "snapshot.storage.k8s.io",
+				Resource: "volumesnapshots",
+			},
+			Namespace: vs.Namespace,
+			Name:      vs.Name,
+		})
+	}
+
+	// Add the VolumeGroupSnapshotClass as an additional item.
+	// Here we assume that the VGS spec defines the class name.
+    // We will be using the default VGSClass for now
+	vgsClassName := vgs.Spec.VolumeGroupSnapshotClassName
+	additionalItems = append(additionalItems, velero.ResourceIdentifier{
+		GroupResource: schema.GroupResource{
+			Group:    "snapshot.storage.k8s.io",
+			Resource: "volumegroupsnapshotclasses",
+		},
+		Namespace: "",
+		Name:      vgsClassName,
+	})
+
+	p.log.Infof("VGS %s ready; adding %d additional items to backup", vgs.Name, len(additionalItems))
+	return item, additionalItems, "", nil, nil
+}
+```
+
+## Implementation
+
+This design proposal is targeted for velero 1.16.
+
+The implementation of this proposed design is targeted for velero 1.17.
+
+## Open Questions
+
+How to handle VGSC ? Do we need VGSC for non-datamover operations ?
```

Comment:
The question has been resolved, right?
The answer is we don't need it.

If so, please remove this line.